### PR TITLE
[SDK][azure] Automatically put requirements.txt to flow.dag.yaml if exists in flow snapshot.

### PR DESCRIPTION
**Pull request by D-W-**
# Description

This pull request includes various changes related to testing, configuration, and code updates for the `Flow` class in the `promptflow` project. The most important changes include adding a new test method to test the behavior of the `_resolve_requirements` method in the `Flow` class, updating the `Flow` class to handle additional configurations and resolve requirements, and adding new constants to the `FlowJobType` class.

Testing improvements:

* [`src/promptflow/tests/sdk_cli_azure_test/unittests/test_flow_entity.py`](diffhunk://#diff-44e987aed05f781a28e90e24fda551a5391751a1d4f59af03a8c1c05cb058720R170-R204): Added a new test method to test the behavior of the `_resolve_requirements` method in the `Flow` class. [[1]](diffhunk://#diff-44e987aed05f781a28e90e24fda551a5391751a1d4f59af03a8c1c05cb058720R170-R204) [[2]](diffhunk://#diff-44e987aed05f781a28e90e24fda551a5391751a1d4f59af03a8c1c05cb058720R16-R17)
* [`src/promptflow/tests/sdk_cli_azure_test/e2etests/test_run_operations.py`](diffhunk://#diff-b63fdccdd1a3a293abe7799c0a231a7226614f726b6de1298f7e0233af26f128R880-R897): Added a new test method that verifies the automatic resolution of requirements when submitting a run. [[1]](diffhunk://#diff-b63fdccdd1a3a293abe7799c0a231a7226614f726b6de1298f7e0233af26f128R880-R897) [[2]](diffhunk://#diff-b63fdccdd1a3a293abe7799c0a231a7226614f726b6de1298f7e0233af26f128R24-R26)

Configuration updates:

* [`src/promptflow/promptflow/azure/_entities/_flow.py`](diffhunk://#diff-b8b37b2a9593a8fb1f0b078663aad36d8b4009a1205bd4e799a6180d666b7e65R123-R140): Updated the `Flow` class to handle additional configurations and resolve requirements. [[1]](diffhunk://#diff-b8b37b2a9593a8fb1f0b078663aad36d8b4009a1205bd4e799a6180d666b7e65R123-R140) [[2]](diffhunk://#diff-b8b37b2a9593a8fb1f0b078663aad36d8b4009a1205bd4e799a6180d666b7e65R28-R29)
* [`src/promptflow/tests/test_configs/flows/flow_with_requirements_txt/flow.dag.yaml`](diffhunk://#diff-88490e1c9edebb1177dc1391010833b47d01e0cb26bb45cdf7c27c64132c62d8R1-R15): Added inputs and outputs to the `flow_with_requirements_txt` flow configuration file.
* [`src/promptflow/tests/test_configs/flows/flow_with_requirements_txt_and_env/print_env.py`](diffhunk://#diff-f65ab3e68fa5947b51697fdc96d7e48d0a30af98d2e33a412918e00cf6fd9274R1-R10): Added a new `print_env.py` code file to the `flow_with_requirements_txt_and_env` flow configuration folder.
* [`src/promptflow/tests/test_configs/flows/flow_with_requirements_txt_and_env/flow.dag.yaml`](diffhunk://#diff-63a2362ed2fea58557c20ca2a3897ce1e85270681847ee59b85c6f3d670669c4R1-R17): Added an `environment` section with `image` set to `python:3.8-slim` in the `flow_with_requirements_txt_and_env` flow configuration file.

Code updates:

* [`src/promptflow/promptflow/azure/_constants/_flow.py`](diffhunk://#diff-556917b04f309689add69a8f5068bda4a5d5c00cae911bfa5685d9383b8c4cd3R25-R27): Added two constants, `ENVIRONMENT` and `PYTHON_REQUIREMENTS_TXT`, to the `FlowJobType` class.
* [`src/promptflow/tests/test_configs/flows/flow_with_requirements_txt/requirements.txt`](diffhunk://#diff-c7ebebd3498172f9293096e8c844f50747317a4923c16044767da343d8d630d5R1): Added the `langchain` package as a requirement to the `requirements.txt` file.
* [`src/promptflow/tests/test_configs/flows/flow_with_requirements_txt/print_env.py`](diffhunk://#diff-e7005605d5d90e28bf123ddca99677fd876e76f9e6f9edb8445fa67d382d3273R1-R13): Updated the `print_env.py` code file to import and print the version of the `langchain` package.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,11 @@
 # Release History
 
+## 1.5.0 (Upcoming)
+
+### Features Added
+
+- [SDK/CLI][azure] Automatically put requirements.txt to flow.dag.yaml if exists in flow snapshot.
```
Comment (wangchao1230) [inline src/promptflow/CHANGELOG.md:7]: 1.4.0 not cut yet, if we can merge today. we can change the change log

File: src/promptflow/promptflow/azure/_entities/_flow.py [hunk 2]
```diff
@@ -78,6 +84,34 @@ def _load_flow_yaml(self, path: Union[str, Path]) -> Dict:
...
+        return True
+
+    @classmethod
+    def _remove_additional_includes(cls, flow_dag: dict):
+        """Remove additional includes from flow dag."""
+        if ADDITIONAL_INCLUDES not in flow_dag:
+            return False
+
+        logger.warning(
+            "Additional includes are found in the flow dag, removing them from flow.dag.yaml after resolved."
```
Comment (wangchao1230) [inline src/promptflow/promptflow/azure/_entities/_flow.py:110]: why this is a warning?

File: src/promptflow/promptflow/azure/_entities/_flow.py [hunk 3]
```diff
@@ -78,6 +84,34 @@ def _load_flow_yaml(self, path: Union[str, Path]) -> Dict:
         """
         return load_yaml(path)
 
+    @classmethod
+    def _resolve_requirements(cls, flow_path: Union[str, Path], flow_dag: dict):
+        """If requirements.txt exists, add it to the flow snapshot. Return if flow_dag is updated."""
```
Comment (wangchao1230) [inline src/promptflow/promptflow/azure/_entities/_flow.py:89]: Return ?? if flow_dag is updated.

File: src/promptflow/promptflow/azure/_entities/_flow.py [hunk 4]
```diff
@@ -78,6 +84,34 @@ def _load_flow_yaml(self, path: Union[str, Path]) -> Dict:
...
+
+        flow_dir = Path(flow_path)
+        if not (flow_dir / cls.DEFAULT_REQUIREMENTS_FILE_NAME).exists():
+            return False
+        if pydash.get(flow_dag, f"{ENVIRONMENT}.{PYTHON_REQUIREMENTS_TXT}"):
+            return False
+        logger.warning(
```
Comment (wangchao1230) [inline src/promptflow/promptflow/azure/_entities/_flow.py:96]: is this an info or warning?

File: src/promptflow/tests/sdk_cli_azure_test/e2etests/test_run_operations.py [hunk 5]
```diff
@@ -875,3 +876,20 @@ def test_get_details_against_partial_completed_run(
...
+
+    def test_auto_resolve_requirements(self, pf: PFClient, randstr: Callable[[str], str]):
+        # will add requirements.txt to flow.dag.yaml if exists when submitting run.
+        run = pf.run(
+            flow=f"{FLOWS_DIR}/flow_with_requirements_txt",
+            data=f"{DATAS_DIR}/env_var_names.jsonl",
+            name=randstr("name"),
+        )
+        pf.runs.stream(run)
+        # # TODO: test snapshot after download can successfully run
```
Comment (wangchao1230) [inline src/promptflow/tests/sdk_cli_azure_test/e2etests/test_run_operations.py]: why commet this?

Comment (D-W-) [reply src/promptflow/tests/sdk_cli_azure_test/e2etests/test_run_operations.py]: Removed in latest change

File: src/promptflow/promptflow/azure/_entities/_flow.py [hunk 6]
```diff
@@ -86,17 +118,24 @@ def _try_build_local_code(self) -> Optional[Code]:
...
+                    dag_updated = self._remove_additional_includes(flow_dag)
                 # promptflow snapshot has specific ignore logic, like it should ignore `.run` by default
-                code._ignore_file = PromptflowIgnoreFile(code.path)
+                code._ignore_file = PromptflowIgnoreFile(flow_dir)
                 # promptflow snapshot will always be uploaded to default storage
                 code.datastore = DEFAULT_STORAGE
+                dag_updated = self._resolve_requirements(flow_dir, flow_dag) or dag_updated
+                if dag_updated:
+                    dump_flow_dag(flow_dag, flow_dir)
             yield code
```
Comment (0mza987) [inline src/promptflow/promptflow/azure/_entities/_flow.py:136]: Better to use try catch here:
```python
  try:
      yield code
  finally:
      if dag_updated:
          dump_flow_dag(original_flow_dag, flow_dir)


```

**Review by wangchao1230 — APPROVED** [review]

**Review by 0mza987 — APPROVED** [review]

**Pull request merged**